-реализовал отображение новостей на вкладке "ИЗБРАННОЕ"

Osip95 opened the pull request.

Osip95 committed:
-реализовал отображение новостей на вкладке "ИЗБРАННОЕ"

Commit:
-создал NewsFragment, для отображения новости по нажатию, добавил в зависимости библиотеку Glide, реализовал открытие нового экрана по нажатию на новость с загрузкой изображения

Commit:
-добавил на экран NewsFragment CoordinatorLayout так, чтобы при скролле изображение плавно с анимацией переходила в toolbar c заголовком новости

Commit:
-добавил на экран NewsFragment кастомный title для toolbar

Commit:
-реализовал открытие экрана NewsFragment с вкладки "ИЗБРАННОЕ",убрал баги в навигации, перешел на один адаптер для двух экранов, добавил toolbar для вкладки "ИЗБРАННОЕ"

ShtefanES commented on app/src/main/java/com/example/newsfeatcherdz/feature/bookmarks/ui/BookmarksFragment.kt:
```diff
@@ -0,0 +1,46 @@
+        val newsFragment = NewsFragment.newInstance(
+            description = viewModel.viewState.value!!.bookmarksArticle[index].description,
+            urlImage = viewModel.viewState.value!!.bookmarksArticle[index].urlToImage,
+            title = viewModel.viewState.value!!.bookmarksArticle[index].title
```
Что-то здесь не то. Если получается так что ты во вью пытаешься обратится к стейту и это не render функция значит скорее всего ты делаешь что-то не то. Вью должна быть максимально тупой, в данном случае этого можно добиться следующими изменениями - сделать так что goNewsEvent излучал ArticleModel и тогда здесь на вью просто возьмешь нужные поля и передашь их  параметром в NewsFragment.newInstance()

ShtefanES commented on app/src/main/java/com/example/newsfeatcherdz/feature/bookmarks/ui/BookmarksFragment.kt:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    private fun render(viewState: ViewState?) {
+        adapter.setData(viewState!!.bookmarksArticle)
```
Вроде можно убрать !! и нулабельность у параметра render

ShtefanES commented on app/src/main/java/com/example/newsfeatcherdz/feature/newsscreen/ui/NewsFragment.kt:
```diff
@@ -0,0 +1,75 @@
+import com.example.newsfeatcherdz.feature.shared.BottomMenuController
+
+
+class NewsFragment : Fragment(R.layout.fragment_news) {
```
Странный нейминг, такое обычно называют как-то вроде ArticleDetailsFragment ну и все остальное по аналогии. Article а не news потому что экран отображает именно сущности Article.

ShtefanES commented on app/src/main/java/com/example/newsfeatcherdz/feature/newsscreen/ui/NewsFragment.kt:
```diff
@@ -0,0 +1,75 @@
+
+        requireActivity().onBackPressedDispatcher.addCallback(this){
+            requireActivity().supportFragmentManager.popBackStack()
+            bottomMenuController.showBottomNavigationMenu()
```
Эти две строки повторяются, хорошо бы в приватную функцию вынести

ShtefanES commented on app/src/main/java/com/example/newsfeatcherdz/feature/newsscreen/ui/NewsFragment.kt:
```diff
@@ -0,0 +1,75 @@
+        val title = arguments?.getString(TITLE_KEY) ?: NOT_FOUND
+
+        newsScreenViewModel =
+            NewsScreenViewModel(title = title, description = description, urlImage = urlImage)
```
Это должно как и другие вью модели через di провайдится. https://insert-koin.io/docs/reference/koin-android/viewmodel/#passing-parameters-to-constructor

В модуле:
    viewModel { parameters ->
        NewsScreenViewModel(title = parameters[0], description = parameters[1], urlImage = parameters[2])
    }
Ну а как передать во фрагменте уже по доке

ShtefanES commented on app/src/main/java/com/example/newsfeatcherdz/feature/shared/ArticlesAdapter.kt:
```diff
@@ -0,0 +1,51 @@
+package com.example.newsfeatcherdz.feature.shared
```
Когда я говорил разместить на том же уровне что feature я имел ввиду именно на одном уровне, те здесь должно быть как-то так:
package com.example.newsfeatcherdz.shared.ui

ShtefanES commented on app/src/main/res/layout/fragment_news.xml:
```diff
@@ -0,0 +1,94 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.coordinatorlayout.widget.CoordinatorLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
На счет этого виджета посмотри вот такой пример https://blog.mindorks.com/using-coordinator-layout-in-android думаю от тебя того же ожидают - что заголовок тоже перемещался, чтоб на экране присутствовала именно стрелочка, это не экран на винде чтоб здесь справа крестик был. В общем это на твое усмотрение, с текущей реализацией скролла может и примут работу как есть, но все же здесь можно сделать гораздо лучше.

ShtefanES commented on app/src/main/res/layout/fragment_bookmarks.xml:
```diff
@@ -0,0 +1,37 @@
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent">
+        <TextView
```
Можешь конечно и так оставить, но в следующей задаче это переделывать нужно будет чтоб все через тему и стили рулилось

ShtefanES commented on app/src/main/res/layout/fragment_bookmarks.xml:
```diff
@@ -0,0 +1,37 @@
+            android:layout_width="wrap_content"
+            android:layout_height="match_parent"
+            android:fontFamily="sans-serif-smallcaps"
+            android:text="@string/app_name"
```
Странно что в заголовок просто название приложение вставляется, обычно если есть вкладки, в заголовке именно название вкладки

Commit:
-изменил функцию NewsFragment.newInstance() и лайфдату goNewsEvent;
- исправил нейминг;
- сделал инициализацию вьмодели экрана ArticleDetailsFragment через di

ShtefanES commented on app/src/main/java/com/example/newsfeatcherdz/feature/article_details_screen/di/module.kt:
```diff
@@ -0,0 +1,16 @@
+package com.example.newsfeatcherdz.feature.article_details_screen.di
```
Хорошо бы переименовать пакет - у тебя часть пакета просто все слитно, а часть через нижнее подчеркивание

ShtefanES commented on app/src/main/java/com/example/newsfeatcherdz/feature/article_details_screen/ui/ArticleDetailsFragment.kt:
```diff
@@ -10,14 +10,21 @@ import androidx.activity.addCallback
-    private lateinit var newsScreenViewModel: NewsScreenViewModel
+    private val articleDetailsScreenViewModel: ArticleDetailsScreenViewModel by viewModel {
+        parametersOf(
+            arguments?.getString(DESCRIPTION_KEY) ?: NOT_FOUND,
```
Обрати внимание какой порядок чтения из параметров в модуле koin и как здесь записываешь данные

ShtefanES commented on app/src/main/java/com/example/newsfeatcherdz/feature/article_details_screen/ui/ArticleDetailsFragment.kt:
```diff
@@ -72,4 +71,9 @@ class NewsFragment : Fragment(R.layout.fragment_news) {
             .into(imgView)
     }
+
+    private fun closeFragment() {
```
Рекомендуется размешать приватные функции под последним использованием

ShtefanES commented on app/src/main/java/com/example/newsfeatcherdz/feature/bookmarks/ui/BookmarksFragment.kt:
```diff
@@ -29,18 +30,18 @@ class BookmarksFragment : Fragment(R.layout.fragment_bookmarks) {
-            description = viewModel.viewState.value!!.bookmarksArticle[index].description,
-            urlImage = viewModel.viewState.value!!.bookmarksArticle[index].urlToImage,
-            title = viewModel.viewState.value!!.bookmarksArticle[index].title
+    private fun goNewsFragment(articleModel: ArticleModel) {
```
Раз фрагмент переиминовался, то этот метод и ивент тоже следует переименовать. Править нужно и на main фрагменте

ShtefanES commented on app/src/main/res/drawable/baseline_west_24.xml:
```diff
@@ -1,5 +1,5 @@
-<vector android:height="24dp" android:tint="#000000"
+<vector android:height="24dp" android:tint="#0A0A0A"
```
Если решил не править поведение с точки зрения ui на деталях новостей, то и менять крестик на стрелку нет смысла, тк стрелка справа плохой UX для пользователей, по гайдам кнопка навигации("стрелочка") располагается слева.

ShtefanES commented on app/src/main/res/values/strings.xml:
```diff
@@ -2,4 +2,6 @@
     <string name="app_name">NewsFeatcherDz</string>
     <string name="main">Главная</string>
     <string name="favorite">Избранное</string>
+    <string name="favorites">Избранное</string>
```
Это ресурс хорошо бы назвать по аналогии с тем что ниже. А вообще почему бы не использовать уже существующие ресурсы выше?

Commit:
пофиксил баг с неправильным выводом текста в ArticleDetailsFragment, перенес метод closeFragment() после последнего использования

Commit:
исправил нейминг

Commit:
сделал поддержку темной темы
